### PR TITLE
AnalyzerAssemblyLoader optimization to search the assemblies loaded b…

### PR DESCRIPTION
**Pull request by mavasani**
…y identity cache

Adrian pointed out that AnalyzerAssemblyLoader.LoadFromPathUnchecked was missing cache hits on _loadedAssembliesByIdentity for same assemblies (analyzer dependencies) loaded from different paths. This was causing large number of redundant ModLoads, which should now be fixed.

Fixes VSO [249557](https://devdiv.visualstudio.com/DevDiv/_workitems?id=249557&fullScreen=false&_a=edit)

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs [hunk 1]
```diff
@@ -53,40 +53,61 @@ public Assembly LoadFromPath(string fullPath)
 
         private Assembly LoadFromPathUnchecked(string fullPath)
         {
+            var assemblyIdentity = AssemblyIdentityUtils.TryGetAssemblyIdentity(fullPath);
+            return LoadFromPathUncheckedCore(fullPath, assemblyIdentity);
+        }
+
+        private Assembly LoadFromPathUncheckedCore(string fullPath, AssemblyIdentity identity)
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: can it be LoadFromPathUnchecked(string, identity = null)? looks like there is only 2 caller of this method.

and do identity = identity ?? AssemblyIdentityUtils.TryGetAssemblyIdentity(fullpath);

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs [hunk 2]
```diff
@@ -53,40 +53,61 @@ public Assembly LoadFromPath(string fullPath)
...
                 if (_loadedAssembliesByPath.TryGetValue(fullPath, out existingAssembly))
                 {
-                    return existingAssembly;
+                    loadedAssembly = existingAssembly;
+                }
+                else if (identity != null && _loadedAssembliesByIdentity.TryGetValue(identity, out existingAssembly))
+                {
+                    loadedAssembly = existingAssembly;
                 }
             }
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: are you missing return loadedAssembly here?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs [hunk 3]
```diff
@@ -53,40 +53,61 @@ public Assembly LoadFromPath(string fullPath)
...
 
-            var identity = AssemblyIdentity.FromAssemblyDefinition(assembly);
-
             lock (_guard)
             {
-                // The same assembly may be loaded from two different full paths (e.g. when loaded from GAC, etc.),
-                // or another thread might have loaded the assembly after we checked above.
-                Assembly existingAssembly;
-                if (_loadedAssembliesByIdentity.TryGetValue(identity, out existingAssembly))
+                if (identity != null)
```
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: this seems either identity = identity ?? AssemlbyIdentity.FromAssemblyDefintiion(assembly) or Contract.ThrowIfNull(identity)

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: Yes, sure.
Additionally, I figured out that the new code will unnecessarily recompute assembly identity from full path for every call to LoadFromPathUnchecked(fullPath). I am going to add another cache for it.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: We need to ensure the same assembly is added to both path and identity cache, AddToCache below does that and returns loaded assembly.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs (hunk 3 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: Yes, that seems better.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs (hunk 2 above)
Comment (heejaechang) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs]: if you change AddCache below, then path and identity cache will always added together I believe.

**Comment by heejaechang** [comment]
:+1:

**Comment by mavasani** [comment]
Tagging @dotnet/roslyn-analysis for review

**Comment by srivatsn** [comment]
Tagging @dotnet/roslyn-compiler

**Comment by mavasani** [comment]
Update from Adrian:

```
I unpacked and built your change on my machine.
The cache is now stable over subsequent analyzer runs i.e. it no longer leaks assemblies if you close and reopen the solution.
I think this will help our customers with long running VS sessions a lot.
```

**Comment by mavasani** [comment]
Ping, any further feedback?

**Comment by tmat** [comment]
:+1:

**Pull request merged**